matrix-sdk/src/androidTest/java/org/matrix/androidsdk/AutoDiscoveryTest.kt

```diff
@@ -0,0 +1,575 @@
+                .build()
+
+        val ad = AutoDiscovery(okHttpClient)
+
+
+        //Act
+        var callbackThread: Thread? = null
+        val lock = CountDownLatch(1)
+        var config: AutoDiscovery.DiscoveredClientConfig? = null
+        ad.findClientConfig("matrix.org", object : TestApiCallback<AutoDiscovery.DiscoveredClientConfig>(lock) {
+            override fun onSuccess(info: AutoDiscovery.DiscoveredClientConfig) {
+                config = info
+                callbackThread = Thread.currentThread()
+                super.onSuccess(info)
+            }
+
+        })
+        lock.await()
+
+
+        //Assert
+        Assert.assertNotNull(config)
+        Assert.assertEquals(AutoDiscovery.Action.FAIL_PROMPT, config!!.action)
+        Assert.assertNull(config?.homeServerUrl)
+        Assert.assertNull(config?.idendityServerUrl)
+        Assert.assertTrue("Callback should be in main thread", Looper.getMainLooper().thread == callbackThread)
+    }
+
+    @Test
+    //  If the response body is empty then FAIL_PROMPT.
+    fun testAutoDiscoveryEmptyBody() {
+
+        //Arrange
+        val mockInterceptor = MockOkHttpInterceptor()
+        mockInterceptor.addRule(MockOkHttpInterceptor.SimpleRule(".well-known/matrix/client", 200, ""))
+
+        val okHttpClient = OkHttpClient.Builder()
+                .addInterceptor(mockInterceptor)
+                .build()
+        val ad = AutoDiscovery(okHttpClient)
+
+        //Act
+        var callbackThread: Thread? = null
+        val lock = CountDownLatch(1)
+        var config: AutoDiscovery.DiscoveredClientConfig? = null
+        ad.findClientConfig("matrix.org", object : TestApiCallback<AutoDiscovery.DiscoveredClientConfig>(lock) {
+            override fun onSuccess(info: AutoDiscovery.DiscoveredClientConfig) {
+                config = info
+                callbackThread = Thread.currentThread()
+                super.onSuccess(info)
+            }
+
+        })
+        lock.await()
+
+        //Assert
+        Assert.assertNotNull(config)
+        Assert.assertEquals(AutoDiscovery.Action.FAIL_PROMPT, config!!.action)
+        Assert.assertNull(config?.homeServerUrl)
+        Assert.assertNull(config?.idendityServerUrl)
+        Assert.assertTrue("Callback should be in main thread", Looper.getMainLooper().thread == callbackThread)
+    }
+
+    @Test
+    //   If the content cannot be parsed, then FAIL_PROMPT.
+    fun testAutoDiscoveryNotJSON() {
+
+        //Arrange
+        val malformedValue = "<html><h1>Hello world!</h1></html>"
+        val mockInterceptor = MockOkHttpInterceptor()
+        mockInterceptor.addRule(MockOkHttpInterceptor.SimpleRule(".well-known/matrix/client", 200, malformedValue))
+
+        val okHttpClient = OkHttpClient.Builder()
+                .addInterceptor(mockInterceptor)
+                .build()
+        val ad = AutoDiscovery(okHttpClient)
+
+        //Act
+        var callbackThread: Thread? = null
```

Comment:
Act is always the same, maybe extract it to a private method?

Reply:
tried but code less readable